EIPS/eip-lock_sload.md

```diff
@@ -0,0 +1,109 @@
+
+And Ethereum's key-value contract storage, unlike traditional databases, _is all about finance_ and highly sensitive balances/amounts. On top of that the extra locking performance overhead doesn't cost anything for a blockchain and its use cases. 
+
+So since the storage is __a database that operates only with sensitive values__ - that's where we need isolation the most. Let's introduce __pessimistic locking by default__ and redefine the SLOAD opcode. That does not break a single valid usage of a contract, yet prevents DAO-like attacks including the DAO itself (__which will be considered a technical fork over a political one from now on__ if this EIP is accepted).
+
+## Specification
+
+Need your advices on this one. I'm sure the idea is outright obvious. The goal is to lock a key that was SLOAD-ed, i.e. turn `select value from Storage0xA where key="Key0xA"` into `select value from Storage0xA where key="Key0xA" SELECT FOR UPDATE`.
+
+Entire Storage0xA locking is probably an overkill. That would kill a few use-cases when the contract is called again in the same frame stack yet operates with different storage keys - no race condition bug there. Instead we should add a global lock array. After each frame termination the locks ackquired in this particular frame must be released.
+
+
+```
+def SLOAD(account, key)
+  lock_id = account+key
+  if $locked.includes(lock_id)
+    throw;
+  else
+    $locked.push(account)
+    old_SLOAD(account, key)
+  end
+end
+
+def RETURN
+  remove last X lock_ids (ones that were added in this frame)
+  old_RETURN
+end
+```
+
+
+## Rationale
+
+Any software (Ethereum is one of them) should make a design change when a specific vulnerability is found way too often. That means no amount of documentation has managed to prevent it and the vulnerability is in underlying platform not in the specific codebase. 
+
+A personal example: a mass assignment vulnerability has been fixed only after it was demonstrated on a high profile website.
+
+That's also what happened to the DAO: the EVM vulnerability was demonstrated (exploited) in a contract that held 14% of internal currency, but unlike Rails, it is not fixed on the underlying platform up to date. As outlined above, it's safe to say it was NOT a vulnerability in the DAO, like many tried to pretend. 
+
+It also certainly does not make it a vulnerability in Solidity. Otherwise, it would be fixed in Solidity already. It is not. What would the fix be? Locks. 
+
+Granular per-record locks inside Solidity would cost way too much gas - each SLOAD would have an overhead of another SLOAD/SSTORE on top of it. 
+
+Per-contract locks would also cost quite some gas and most contracts would opt-out from it. 
+
+## Backwards Compatibility
+
+The behavior should be changed from a certain block height. We should run a test how many transactions throughout entire history would throw with new consensus (wild guess: only the DAO one)
```

Comment:
My guess is like 20 more contracts would be affected.  I'm interested in the result.